models/user.go

```diff
@@ -1354,6 +1355,113 @@ func GetWatchedRepos(userID int64, private bool) ([]*Repository, error) {
 	return repos, nil
 }
 
+// deleteKeysMarkedForDeletion returns true if ssh keys needs update
+func deleteKeysMarkedForDeletion(keys []string) (sshKeysNeedUpdate bool, err error) {
+	// Start session
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return false, err
+	}
+
+	// Delete keys marked for deletion
+	for _, KeyToDelete := range keys {
+		key, err := SearchPublicKeyByContent(KeyToDelete)
+		err = deletePublicKeys(sess, key.ID)
+		if err != nil {
+			sshKeysNeedUpdate = false
+			log.Trace("deleteKeysMarkedForDeletion: Error: %v", key)
+		}
+		sshKeysNeedUpdate = true
+	}
+
+	return sshKeysNeedUpdate, sess.Commit()
+}
+
+func addLdapSSHPublicKeys(s *LoginSource, usr *User, SSHPublicKeys []string) (sshKeysNeedUpdate bool, err error) {
+	for _, LDAPPublicSSHKey := range SSHPublicKeys {
+		if strings.HasPrefix(strings.ToLower(LDAPPublicSSHKey), "ssh") {
+			LDAPPublicSSHKeyName := strings.Join([]string{s.Name, LDAPPublicSSHKey[0:40]}, "-")
+			_, err := AddPublicKey(usr.ID, LDAPPublicSSHKeyName, LDAPPublicSSHKey, s.ID)
+			if err != nil {
+				log.Error(4, "addLdapSSHPublicKeys[%s]: Error adding LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, err)
+			} else {
+				log.Trace("addLdapSSHPublicKeys[%s]: Added LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, LDAPPublicSSHKey)
+				sshKeysNeedUpdate = true
+			}
+		} else {
+			log.Error(3, "addLdapSSHPublicKeys[%s]: Skipping invalid LDAP Public SSH Key for user %s: %v", s.Name, usr.Name, LDAPPublicSSHKey)
+		}
+	}
+	return sshKeysNeedUpdate, err
+}
+
+func synchronizeLdapSSHPublicKeys(s *LoginSource, SSHPublicKeys []string, usr *User) (sshKeysNeedUpdate bool, err error) {
+	log.Trace("synchronizeLdapSSHPublicKeys[%s]: Handling LDAP Public SSH Key synchronization for user %s", s.Name, usr.Name)
+
+	// Get Public Keys from DB with LDAP source
+	var giteaKeys []string
+	keys, err := ListPublicKeys(usr.ID)
```

Comment:
Why not just only load all public keys where login_source_id > 0?

Reply:
Sure, that would be more efficient. Would you recommend modifying the current ListPublicKeys for this purpose (adding optional 'where login_source_id>x'), creating a new function or any other method to accomplish that?

Reply:
@dnmgns Yes, I think you should do that for performance.